pkg/cache/v3/linear.go

```diff
@@ -114,60 +114,116 @@ func NewLinearCache(typeURL string, opts ...LinearCacheOption) *LinearCache {
-		for _, resource := range cache.resources {
-			resources = append(resources, types.ResourceWithTTL{Resource: resource})
+func (cache *LinearCache) computeSotwResponse(watch ResponseWatch) *RawResponse {
+	var changedResources []string
+	var removedResources []string
+	knownVersions := watch.subscription.ReturnedResources()
+
+	if watch.subscription.IsWildcard() {
+		for resourceName, version := range cache.versionVector {
+			knownVersion, ok := knownVersions[resourceName]
+			if !ok {
+				// This resource is not yet known by the client (added or the client has a stale view)
+				changedResources = append(changedResources, resourceName)
+			} else if knownVersion != version {
+				// Outdated version
+				changedResources = append(changedResources, resourceName)
+			}
 		}
-	} else if ResourceRequiresFullStateInSotw(cache.typeURL) {
-		// Non-wildcard request for a type requiring full state response
-		// We need to return all requested resources, if existing, for this type
-		requestedResources := watch.Request.GetResourceNames()
-		resources = make([]types.ResourceWithTTL, 0, len(requestedResources))
-		for _, resource := range requestedResources {
-			resource := cache.resources[resource]
-			if resource != nil {
-				resources = append(resources, types.ResourceWithTTL{Resource: resource})
+
+		// Negative check to retrieve resources that would have been removed
+		for resourceName := range knownVersions {
+			if _, ok := cache.versionVector[resourceName]; !ok {
+				// Known resource no longer exists.
+				removedResources = append(removedResources, resourceName)
 			}
 		}
 	} else {
-		// Non-wildcard request for other types. Only return stale resources
-		resources = make([]types.ResourceWithTTL, 0, len(staleResources))
-		for _, name := range staleResources {
-			resource := cache.resources[name]
-			if resource != nil {
-				resources = append(resources, types.ResourceWithTTL{Resource: resource})
+		for resourceName := range watch.subscription.SubscribedResources() {
+			version, exists := cache.versionVector[resourceName]
+			knownVersion, known := knownVersions[resourceName]
+			if !exists {
+				if known {
+					// This resource was removed and is no longer known. If the type requires full state
+					// we need to return to make it aware
+					removedResources = append(removedResources, resourceName)
+				}
+				continue
+			}
+
+			if !known {
+				// This resource is not yet known by the client (added or the client has a stale view)
+				changedResources = append(changedResources, resourceName)
+			} else if knownVersion != version {
+				// Outdated version
+				changedResources = append(changedResources, resourceName)
 			}
 		}
 	}
-	watch.Response <- &RawResponse{
-		Request:   watch.Request,
-		Resources: resources,
-		Version:   cache.getVersion(),
-		Ctx:       context.Background(),
+
+	if len(changedResources) == 0 && len(removedResources) == 0 {
+		// Nothing changed
+		return nil
```

Comment:
There are multiple cases when this can happen:
 - ACK
 - Request is adding resources not present in the cache
 - Request is removing subscription
 - probably others